### PR TITLE
!str #16951 Unify OperationAttributes, and move actor specifics ActorOperationAttributes

### PR DESCRIPTION
**Pull request by patriknw**
Refs #16951

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/OperationAttributes.scala [hunk 1]
```diff
@@ -33,9 +43,9 @@ final case class OperationAttributes private (attributes: List[OperationAttribut
   /**
    * INTERNAL API
    */
-  private[akka] def name: String = nameLifted match {
+  private[akka] def nameOrDefault(default: String = "unknown-operation"): String = nameLifted match {
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/OperationAttributes.scala]: renamed due to overload issue from java

File: akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala [hunk 2]
```diff
@@ -75,7 +77,7 @@ class FlowErrorDocSpec extends AkkaSpec {
         if (elem < 0) throw new IllegalArgumentException("negative not allowed")
         else acc + elem
       }
-      .withAttributes(OperationAttributes.supervisionStrategy(decider))
+      .withAttributes(ActorOperationAttributes.supervisionStrategy(decider))
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala]: +1

File: akka-stream/src/main/scala/akka/stream/OperationAttributes.scala [hunk 3]
```diff
@@ -44,21 +54,28 @@ final case class OperationAttributes private (attributes: List[OperationAttribut
...
 
 }
 
+/**
+ * Note that more attributes for the [[ActorFlowMaterializer]] are defined in [[ActorOperationAttributes]].
+ */
 object OperationAttributes {
 
-  sealed trait Attribute
+  trait Attribute
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/OperationAttributes.scala]: Great, so now they're user extendable

**Comment by ktoso** [comment]
LGTM

File: akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala (hunk 2 above)
Comment (drewhk) [inline akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala]: Can we abbreviate? ActorAttributes? Maybe we should rename OperationAttributes, too.

File: akka-stream/src/main/scala/akka/stream/OperationAttributes.scala [hunk 4]
```diff
@@ -1,20 +1,30 @@
...
+ * Note that more attributes for the [[ActorFlowMaterializer]] are defined in [[ActorOperationAttributes]].
  */
 final case class OperationAttributes private (attributes: List[OperationAttributes.Attribute] = Nil) {
 
   import OperationAttributes._
 
   /**
+   * Java API
+   */
+  def getAttributes(): java.util.List[Attribute] = {
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/OperationAttributes.scala]: Maybe have a class keyed version that filters on a given class?

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
akka-docs-java-lambda failed to build AFAICS

File: akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala (hunk 2 above)
Comment (patriknw) [inline akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala]: yes, perhaps `Attributes` and `ActorAttributes` are enough

Comment (rkuhn) [inline akka-docs-dev/rst/scala/code/docs/stream/FlowErrorDocSpec.scala]: Wouldn’t the user typically use static imports for the members anyway?

File: akka-stream/src/main/scala/akka/stream/OperationAttributes.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/OperationAttributes.scala]: good point; probably both—filtered list and collectFirst option

File: akka-stream/src/main/scala/akka/stream/OperationAttributes.scala [hunk 5]
```diff
@@ -76,14 +93,31 @@ object OperationAttributes {
...
+   * Scala API: Decides how exceptions from user are to be handled.
    */
   def supervisionStrategy(decider: Supervision.Decider): OperationAttributes =
     OperationAttributes(SupervisionStrategy(decider))
+
+  /**
+   * Java API: Decides how exceptions from application code are to be handled.
+   */
+  def withSupervisionStrategy(decider: japi.Function[Throwable, Supervision.Directive]): OperationAttributes =
+    ActorOperationAttributes.supervisionStrategy(e ⇒ decider.apply(e))
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/OperationAttributes.scala]: for these I prefer eta expansion syntax: `(decider.apply _)`

File: akka-stream/src/main/scala/akka/stream/OperationAttributes.scala (hunk 4 above)
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/OperationAttributes.scala]: actually scratch the second one: we should be more opinionated and offer these two
~~~scala
def getAttributes[T <: Attribute](c: Class[T]): java.util.List[T]
def getAttribute[T <: Attribute](c: Class[T], default: T): T
~~~

**Comment by rkuhn** [comment]
LGTM, fit for purpose already, but might as well add the last bits of user-extensibility helpers I commented on.

**Comment by ktoso** [comment]
@patriknw I'd love to get this one in today so I can rebase my other PRs :-) 
Seems only Roland's usability improvement suggestion for Java API is missing.

**Comment by patriknw** [comment]
alright, I can need a break from the other thing anyway, I'll complete this now

**Comment by patriknw** [comment]
added the convenience methods, and squashed
will merge after validation

**Comment by ktoso** [comment]
LGTM, thanks for finishing it up :-)

**Pull request merged**